File: trunk/Java/OPSJLib/src/ops/Subscriber.java
/**
*
* Copyright (C) 2006-2009 Anton Gravestam.
*
* This file is part of OPS (Open Publish Subscribe).
*
* OPS (Open Publish Subscribe) is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.

* OPS (Open Publish Subscribe) is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public License
* along with OPS (Open Publish Subscribe).  If not, see <http://www.gnu.org/licenses/>.
*/
package ops;

import java.util.ArrayList;
import java.util.Observable;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.ReentrantLock;
import ops.protocol.OPSMessage;

/**
 *
 * @author Anton Gravestam
 */
public class Subscriber extends Observable 
{
    public Event deadlineEvent = new Event();

    private Topic topic;
    private String identity = "";
    private ArrayList<FilterQoSPolicy> filterQoSPolicies = new ArrayList<FilterQoSPolicy>();
    private MessageFilterSet messageFilters = new MessageFilterSet();
    private final Object newDataEvent = new Object();
    private OPSObject data;
    private long deadlineTimeout;
    private long lastDeadlineTime;
    private long timeLastDataForTimeBase;
    private long timeBaseMinSeparationTime;
    private boolean reliable;
    private ReceiveDataHandler receiveDataHandler;
    protected Participant participant;
    private OPSMessage message;
    private final DeadlineNotifier deadlineNotifier;

    public Subscriber(Topic t)
    {
        this.topic = t;
        this.participant = Participant.getInstance(topic.getDomainID(), topic.getParticipantID());
        receiveDataHandler = participant.getReceiveDataHandler(t);
        deadlineNotifier = DeadlineNotifier.getInstance();
    }

    public synchronized void setDeadlineQoS(long timeout)
    {
        deadlineTimeout = timeout;
    }

    public synchronized void setTimeBasedFilterQoS(long minSeparationTime)
    {
        timeBaseMinSeparationTime = minSeparationTime;
    }

    public synchronized void start()
    {

        lastDeadlineTime = System.currentTimeMillis();
        timeLastDataForTimeBase = System.currentTimeMillis();

        deadlineNotifier.add(this);
        receiveDataHandler.addSubscriber(this);

    }

    public synchronized boolean stop()
    {
        deadlineNotifier.remove(this);
        return receiveDataHandler.removeSubscriber(this);
    }
    public synchronized boolean isDeadlineMissed()
    {
        if (System.currentTimeMillis() - lastDeadlineTime > deadlineTimeout)
        {
            return true;
        }
        return false;
    }

    protected synchronized void checkDeadline()
    {
        if (System.currentTimeMillis() - lastDeadlineTime > deadlineTimeout)
        {
            deadlineEvent.fireEvent();
            lastDeadlineTime = System.currentTimeMillis();
        }

    }

    public synchronized boolean isReliable()
    {
        return reliable;
    }

    public synchronized String getIdentity()
    {
        return identity;
    }

    protected synchronized void notifyNewOPSObject(OPSObject o)
    {

        if (applyFilterQoSPolicies(o))
        {
            if (System.currentTimeMillis() - timeLastDataForTimeBase > timeBaseMinSeparationTime || timeBaseMinSeparationTime == 0)
            {
                lastDeadlineTime = System.currentTimeMillis();
                timeLastDataForTimeBase = System.currentTimeMillis();
                setChanged();
                data = o;//(OPSObject) o.clone();
                notifyObservers(data);
                synchronized(newDataEvent)
                {
                    newDataEvent.notifyAll();
                }

            }
        }
    }

    public OPSObject waitForNextData(long millis)
    {
        try
        {
            synchronized(newDataEvent)
            {
                newDataEvent.wait(millis);
            }
            return data;
        }
        catch (InterruptedException ex)
        {
            return null;
        }
       
    }

    public synchronized void addFilterQoSPolicy(FilterQoSPolicy qosPolicy)
    {
        synchronized (this)
        {
            getFilterQoSPolicies().add(qosPolicy);
        }
    }

    public synchronized void removeFilterQoSPolicy(FilterQoSPolicy qosPolicy)
    {
        synchronized (this)
        {
            getFilterQoSPolicies().remove(qosPolicy);
        }
    }

    synchronized void notifyNewOPSMessage(OPSMessage message)
    {
        if(messageFilters.applyFilter(message))
        {
            this.message = message;
            notifyNewOPSObject(message.getData());
        }
    }

    private boolean applyFilterQoSPolicies(OPSObject o)
    {
        //throw new UnsupportedOperationException("Not yet implemented");
        for (FilterQoSPolicy filter : filterQoSPolicies)
        {
            if (!filter.applyFilter(o))
            {
                //Indicates that this data sample NOT should be delivered to the application.
                return false;
            }

        }
        return true;
    }

    public synchronized ArrayList<FilterQoSPolicy> getFilterQoSPolicies()
    {
        return filterQoSPolicies;
    }

    public synchronized OPSMessage getMessage()
    {
        return message;
    }
    

    protected synchronized OPSObject getData()
    {
        return data;
    }

    public synchronized void removeFilter(MessageFilter filter)
    {
        messageFilters.removeFilter(filter);
    }

    public synchronized void addFilter(MessageFilter filter)
    {
        messageFilters.addFilter(filter);
    }

    public synchronized MessageFilterSet getMessageFilters()
    {
        return messageFilters;
    }





}

File: trunk/Cpp/source/Subscriber.cpp
/**
 *
 * Copyright (C) 2006-2009 Anton Gravestam.
 *
 * This file is part of OPS (Open Publish Subscribe).
 *
 * OPS (Open Publish Subscribe) is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * OPS (Open Publish Subscribe) is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with OPS (Open Publish Subscribe).  If not, see <http://www.gnu.org/licenses/>.
 */
#include "OPSTypeDefs.h"
#include "Subscriber.h"
#include "TimeHelper.h"
#include "Participant.h"
#include <boost/thread/mutex.hpp>
#include <boost/thread/condition_variable.hpp>
#include "boost/date_time/local_time/local_time.hpp"


namespace ops
{

    Subscriber::Subscriber(Topic t) :
    topic(t),
    hasUnreadData(false),
    deadlineTimeout(TimeHelper::infinite),
    timeBaseMinSeparationTime(0),
    threadPolicy(SHARED_THREAD),
    deadlineMissed(false),
    firstDataReceived(false),
    messageBufferMaxSize(1),
    started(false)
    //deadlineTimer(*Participant::getIOService())
    {
        message = NULL;
        data = NULL;

        participant = Participant::getInstance(topic.getDomainID(), topic.getParticipantID());
        //participant = topic.getParticipant();
        deadlineTimer = DeadlineTimer::create(participant->getIOService());
        
        newDataMutex = new boost::mutex();
        newDataEvent = new boost::condition_variable;//CreateEvent(NULL, true, false, NULL);
        timeLastData = TimeHelper::currentTimeMillis();



    }

    Subscriber::~Subscriber()
    {
        listeners.clear();
        if (started)
        {
            stop();
        }
        delete deadlineTimer;

		while (messageBuffer.size() > 0) {
			messageBuffer.back()->unreserve();
			messageBuffer.pop_back();
		}
		delete newDataMutex;
		delete newDataEvent;
    }

    void Subscriber::start()
    {
		if (started) return;

        receiveDataHandler = participant->getReceiveDataHandler(topic);
        receiveDataHandler->addListener(this);
        deadlineTimer->addListener(this);
        deadlineTimer->start(deadlineTimeout);
        started = true;
    }

    void Subscriber::stop()
    {
		if (!started) return;

		receiveDataHandler->aquireMessageLock();
        receiveDataHandler->removeListener(this);
        receiveDataHandler->releaseMessageLock();
		receiveDataHandler = NULL;
        participant->releaseReceiveDataHandler(topic);
        deadlineTimer->removeListener(this);
        deadlineTimer->cancel();
        started = false;
    }

    void Subscriber::onNewEvent(Notifier<OPSMessage*>* sender, OPSMessage* message)
    {
        //Perform a number of checks on incomming data to be sure we want to deliver it to the application layer

        //Check that this message is delivered on the same topic as this Subscriber use
        if (message->getTopicName() != topic.getName())
        {
            return;
        }
        //Check that the type of the delivered data can be interpreted as the type we expect in this Subscriber
        else if (message->getData()->getTypeString().find(topic.getTypeID()) == std::string::npos)
        {
            return;
        }

        //OK, we passed the basic checks, lets go on and filter on data content...

        OPSObject* o = message->getData();
        if (applyFilterQoSPolicies(o))
        {
            if (TimeHelper::currentTimeMillis() - timeLastDataForTimeBase > timeBaseMinSeparationTime || timeBaseMinSeparationTime == 0)
            {
                firstDataReceived = true;
                
                //saveCopy(o);
                addToBuffer(message);
                this->message = message;
                data = o;

                notifyNewData();

                boost::unique_lock<boost::mutex> lock(*newDataMutex);
                hasUnreadData = true;
                newDataEvent->notify_all();
                lock.unlock();

                timeLastDataForTimeBase = TimeHelper::currentTimeMillis();
                timeLastData = TimeHelper::currentTimeMillis();
                setDeadlineMissed(false);

                deadlineTimer->start(deadlineTimeout);

                //cancelDeadlineTimeouts();
                //registerForDeadlineTimeouts();

            }
        }
    }

    void Subscriber::addToBuffer(OPSMessage* mess)
    {
        mess->reserve();
        messageBuffer.push_front(mess);
        while (messageBuffer.size() > messageBufferMaxSize)
        {
            messageBuffer.back()->unreserve();
            messageBuffer.pop_back();
        }
    }

    void Subscriber::setHistoryMaxSize(int s)
    {
        messageBufferMaxSize = s;

    }

    std::deque<OPSMessage*> Subscriber::getHistory()
    {
        return messageBuffer;
    }

    OPSObject* Subscriber::getData()
    {
        hasUnreadData = false;
        return data;
    }

    void Subscriber::setData(OPSObject* data)
    {
        SafeLock lock(this);
        this->data = data;
    }

    Topic Subscriber::getTopic()
    {
        return topic;
    }

    void Subscriber::addFilterQoSPolicy(FilterQoSPolicy* fqos)
    {
        //WaitForSingleObject(filterQoSPolicyMutex, INFINITE);
        SafeLock lock(&filterQoSPolicyMutex);
        filterQoSPolicies.push_back(fqos);
        
    }

    void Subscriber::removeFilterQoSPolicy(FilterQoSPolicy* fqos)
    {
        SafeLock lock(&filterQoSPolicyMutex);
        filterQoSPolicies.remove(fqos);
        
    }

    bool Subscriber::applyFilterQoSPolicies(OPSObject* obj)
    {
        SafeLock lock(&filterQoSPolicyMutex);
        bool ret = true;
        std::list<FilterQoSPolicy*>::iterator p;
        p = filterQoSPolicies.begin();
        while (p != filterQoSPolicies.end())
        {
            if (!(*p)->applyFilter(obj))
            {
                ret = false;
                break;
            }
            p++;
        }
       
        return ret;

    }

    void Subscriber::setDeadlineQoS(__int64 millis)
    {
		if (millis == 0) {
		    deadlineTimeout = TimeHelper::infinite;
		} else {
	        deadlineTimeout = millis;
		}
		cancelDeadlineTimeouts();	// Restart with new timeout
    }

    __int64 Subscriber::getDeadlineQoS()
    {
        return deadlineTimeout;
    }

    void Subscriber::checkAndNotifyDeadlineMissed()
    {
        if (isDeadlineMissed())
        {
            //printf("DeadlineMissed timeLastData = %d, currTime = %d, deadlineTimeout = %d\n", timeLastData, currTime, deadlineTimeout);
            deadlineMissedEvent.notifyDeadlineMissed();
            timeLastData = TimeHelper::currentTimeMillis();
        }

    }

    __int64 Subscriber::getTimeBasedFilterQoS()
    {
        return timeBaseMinSeparationTime;
    }

    void Subscriber::setTimeBasedFilterQoS(__int64 timeBaseMinSeparationMillis)
    {
        timeBaseMinSeparationTime = timeBaseMinSeparationMillis;
    }

    bool Subscriber::waitForNewData(int timeout)
    {
        if (hasUnreadData)
        {
            return true;
        }

        boost::unique_lock<boost::mutex> lock(*newDataMutex);

        boost::system_time time = boost::get_system_time();
        time += boost::posix_time::milliseconds(timeout);

        if(newDataEvent->timed_wait(lock, time))
        {
            return true;
        }
        return false;
//        else if (WaitForSingleObject(newDataEvent, timeout) != WAIT_TIMEOUT)
//        {
//            ResetEvent(newDataEvent);
//            return true;
//        }
//        else
//        {
//            return false;
//        }
    }

    int Subscriber::getThreadPolicy()
    {
        return threadPolicy;
    }

    void Subscriber::setThreadPolicy(int threadPolicy)
    {
        this->threadPolicy = threadPolicy;
    }

    std::string Subscriber::getName()
    {
        return name;
    }

    void Subscriber::setName(std::string name)
    {
        this->name = name;
    }

    bool Subscriber::isDeadlineMissed()
    {
        __int64 currTime = TimeHelper::currentTimeMillis();
        if (currTime - timeLastData > deadlineTimeout)
        {
            deadlineMissed = true;
            return deadlineMissed;
        }
        return false;
    }

    void Subscriber::setDeadlineMissed(bool deadlineMissed)
    {
        this->deadlineMissed = deadlineMissed;
    }

    void Subscriber::registerForDeadlineTimeouts()
    {
        deadlineTimer->addListener(this);
        //boost::asio::deadline_timer t(Participant::getIOService(), boost::posix_time::milliseconds(deadlineTimeout));
        //deadlineTimer.cancel();

        /*if(deadlineTimer.expires_from_now(boost::posix_time::milliseconds(deadlineTimeout)) > 0)
        {*/
        //deadlineTimer.async_wait(boost::bind(&Subscriber::asynchHandleDeadlineTimeout, this, boost::asio::placeholders::error));
        //}

    }

    void Subscriber::cancelDeadlineTimeouts()
    {
        deadlineTimer->start(deadlineTimeout);
        //if(deadlineTimer.expires_from_now(boost::posix_time::milliseconds(deadlineTimeout)) > 0)
        //{
        //	//deadlineTimer.async_wait(boost::bind(&Subscriber::asynchHandleDeadlineTimeout, this, boost::asio::placeholders::error));
        //}
        //else
        //{
        //	/*boost::asio::deadline_timer newTimer(*Participant::getIOService());
        //	deadlineTimer = newTimer;
        //	deadlineTimer.expires_from_now(boost::posix_time::milliseconds(deadlineTimeout));*/
        //}

    }
    //void Subscriber::asynchHandleDeadlineTimeout(const boost::system::error_code& e)
    //{
    //	if (e != boost::asio::error::operation_aborted)
    //	{
    //		// Timer was not cancelled, take necessary action.
    //		deadlineMissedEvent.notifyDeadlineMissed();

    //		cancelDeadlineTimeouts();
    //		registerForDeadlineTimeouts();
    //	}
    //
    //	//t.reset();
    //
    //}

    void Subscriber::onNewEvent(Notifier<int>* sender, int message)
    {
        deadlineMissedEvent.notifyDeadlineMissed();
        deadlineTimer->start(deadlineTimeout);
        //cancelDeadlineTimeouts();
    }

    bool Subscriber::aquireMessageLock()
    {
        return receiveDataHandler->aquireMessageLock();
    }

    void Subscriber::releaseMessageLock()
    {
        receiveDataHandler->releaseMessageLock();
    }

    OPSMessage* Subscriber::getMessage()
    {
        return message;
    }


}
